src/create.rs

```diff
@@ -461,12 +459,24 @@ impl Execution {
                 self.capture_output,
             ));
         }
+        let wixobj_sources = self.wixobj_sources(&wixobj_destination)?;
```

Comment:
I think all of this can be simplified to as follows:

```rust
let is_bundle = wixobj_sources
    .iter()
    .map(InstallerKind::try_from)
    .collect::<Result<Vec<InstallerKind>>>()?
    .contains(&InstallerKind::Bundle);
debug!("is_bundle = {:?}", is_bundle);
let msi_destination = self.msi_destination(
        &name,
        &version,
        platform,
        debug_name,
        is_bundle,
        &manifest,
)?;
```

or

```rust
let is_bundle = wixobj_sources
    .iter()
    .map(InstallerKind::try_from)
    .collect::<Result<Vec<InstallerKind>>>()?
    .iter()
    .any(InstallerKind::is_bundle);
debug!("is_bundle = {:?}", is_bundle);
let msi_destination = self.msi_destination(
        &name,
        &version,
        platform,
        debug_name,
        is_bundle,
        &manifest,
)?;
```

which would eliminate the need for the `std::ops::Add`, `std::ops::AddAssign`, and `std::iter::Sum` trait implementations on the `InstallerKind` enum type. I think the combination of these implementations is essentially the same thing as using the `contains` or `any` methods. The `collect` is needed for the error handling.

I think this would also eliminate the need for the `InstallerKind::None` and `InstallerKind::Both` variants, which appears to only be used in initializer for the `fold` method within the `sum` implementation and the `add` method implementation, respectively, but I might be missing something here. 

I tried both types on my local copy of the PR, and it all compiles, but I have not tested it. Do you happen to have a test bundle-style WXS files to use?

Reply:
The goal is to be able to support more than just "is bundle".  For example, how does the linker react when both `product` and `bundle` intermediate files are included?  If it reacts poorly having `InstallerKind::Both` makes it trivial to skip linking / output a reasonable message.

Reply:
The data provided by the "rolled up" value just helped me troubleshoot a build pipeline problem.  I advocate to keep the code as-is.

Reply:
Okay, I thought I was following the code, but now I am a little lost.

You iterate through all WIXOBJ files and determine their respective installer kind and then "reduce" all installer kinds down to a single value. This is essentially the purpose of the `fold` method. From the [`fold` method documentation](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold):

>An iterator method that applies a function, producing a single, final value.
> ...
>This operation is sometimes called 'reduce' or 'inject'.
>
>Folding is useful whenever you have a collection of something, and want to produce a single value from it.

Why wrap the `fold` method into a `sum` method, which requires implementing `add`? Installer kinds are not numeric values, so "summing" does not make semantic sense to me and seems to complicate the implementation. I am having a hard time wrapping by brain around this implementation because of the "summing" and "adding". Couldn't you achieve the same thing with just using the `fold` method directly:

```rust
let is_bundle = wixobj_sources
    .iter()
    .map(InstallerKind::try_from)
    .collect::<Result<Vec<InstallerKind>>>()?
    .iter()
    .fold(InstallerKind::None, |a, v| {
        // I chose to change the `!=` around as it was confusing me. I try to make the
        // if block in a if-else statement the affirmative case by convention.
        if v == InstallerKind::Unknown {
            a
        } else {
            if a == v {
                a
            } else if a == InstallerKind::None {
                v
            } else if v == InstallerKind::None {
                a
            } else {
                InstallerKind::Both
            }
        }
    }).is_bundle();
```

This would maintain the same functionality, but be more meaningful (to current-me and possibly future-me). I guess I am just getting stuck with the `add` and `sum` implementations. It is weird to me to do arithmetic on an ultimately non-numeric type.

>The goal is to be able to support more than just "is bundle". For example, how does the linker react when both product and bundle intermediate files are included? If it reacts poorly having InstallerKind::Both makes it trivial to skip linking / output a reasonable message.

I was under the impression that bundles contain products and the existence of at least one bundle intermediate file indicates the installer is to be a bundle/exe instead of a product/msi. If this is true and both types of intermediate files exist, then the linker reacts by creating a bundle/exe destination. 

If the installer kind is `InstallerKind::Both`, the `is_bundle` returns `false`, a `.msi` file name is used, but the bundle intermediate file is used by the linker and it complains, correct? As the user of the cargo-wix subcommand, how do I go about fixing this issue? I appear to be missing a project layout/configuration that uses bundles and products that will cause an error in the linker not related to the file extension.

>The data provided by the "rolled up" value just helped me troubleshoot a build pipeline problem. I advocate to keep the code as-is.

I am fine keeping the `InstallerKind::Both` variant, especially if it helps others debug their projects, but I am just trying to wrap my head around the implementation and "wondering out loud" if there is an alternative. However, I think it is possible to remove the `InstallerKind::None` variant with the following fold-only implementation from above:

```rust
let is_bundle = wixobj_sources
    .iter()
    .map(InstallerKind::try_from)
    .collect::<Result<Vec<InstallerKind>>>()?
    .iter()
    .fold(InstallerKind::Unknown, |a, v| {
        if v == InstallerKind::Unknown {
            a
        } else if a == v {
            a
        } else {
            InstallerKind::Both
        }
    }).is_bundle();
```

Basically, I think Unknown can be used as the initializer because the installer kind is unknown at the start. It is not "none". It is something, but we just don't know it yet. If all WIXOBJs are products, this should yield a `InstallerKind::Product`, and likewise, if all WIXOBJs are bundles, this should yield a `InstallerKind::Bundle`. At any point if there is a mismatch in kinds, then this should return `InstallerKind::Both`. This maintains the same functionality.

Another thought, which you may hate me for, `InstallerKind` does not fit the current usage. This should be `ObjectKind` or `WixObjKind` because it is ultimately the kind of intermediate file not the kind of installer. An `InstallerKind` would be either `Unknown`, `Msi`, or `Exe`. I am not sure how this fits into the `Both` variant because an intermediate file cannot be a product and bundle at the same, can it? I guess the `Both` variant would translate to the `Unknown` variant because we cannot determine the kind of installer...hmm, `Indeterminate` variant? This would requiring mapping `WixObjKind` to an `InstallerKind` instead of reducing `WixObjKind` to a single value. Note, this thought will not block merging of the PR.